### ISSUE TITLE
core dumped when import hnswlib

### ISSUE THREAD

**Issue by universewill** [nmslib/hnsw#17]
The error also occurs when running the example code of hnswlib. It seems it is correlated with the system eviroment, but i can not figure it out. Same code runs well on another machine.
The error occurs when running this code line:
```
p.init_index(max_elements = num_elements, ef_construction = 200, M = 16)
```

I reinstall unbuntu system and hnsw, and it runs into core dump when import hnswlib ...

Any hint or suggestions here?

**Comment by yurymalkov** [comment]
@universewill thanks!
```import hnswlib``` precedes any operations with the index.
It is probably either something is wrong with pybind (in this case the behavior would be different for different versions) or cpu-instructions related stuff.
What is the cpu model?

**Comment by universewill** [comment]
![fail_cpu](https://user-images.githubusercontent.com/6790730/40904021-f71b909a-680b-11e8-9f2b-34d8f4a0bc07.png)

**Comment by yurymalkov** [comment]
@universewill Ok. Thanks!
I think I see what is the problem. I'll try to make a fix.

**Comment by universewill** [comment]
I test the fix-avx package, and it runs well. 
Could you explain the reason of the error?
Many thanks! @yurymalkov

**Comment by yurymalkov** [comment]
@universewill Thanks!
That was a stupid error (old hardcode) which lead to absence of compile time AVX checks. 
All of the computers that I use support AVX, so I didn't notice it.

**Comment by zyh3826** [comment]
Hi, I got the same error. this url:  https://github.com/nmslib/hnsw/tree/fix-avx is invalid, could you please post the right url? @yurymalkov

**Comment by yurymalkov** [comment]
Hi @zyh3826 The branch has been merged to master a long time ago. I wonder, how did you install hnswlib?

**Comment by zyh3826** [comment]
Hi @yurymalkov thanks for your reply, I install hnswlib by:
> git clone https://github.com/nmslib/hnswlib.git
cd hnswlib
pip install .

**Comment by yurymalkov** [comment]
Hm. It seems you have a separate problem. Can you share your code?

**Comment by zyh3826** [comment]
@yurymalkov my code:

> hnsw_index = hnswlib.Index(space="angulardist", dim=200)
hnsw_index.init_index(max_elements=8800000, ef_construction=100, M=128)
hnsw_index.add_item(embeddings.numpy(), list(vocab.values()))
hnsw_index.save_index(model_save_path)

embeddings is a torch.FloatTensor
the error occurs at hnsw_index.init_index

**Comment by yurymalkov** [comment]
@zyh3826 Hm. Are you sure you have enough RAM to fit the dataset? What happens if, say, `max_elements= 1000`?

**Comment by zyh3826** [comment]
@yurymalkov my RAM is about 260Gb. Is this enough?

**Comment by yurymalkov** [comment]
That depends on many factors, but such hnswlib index should not take more than 30 Gb.

**Comment by jdthorpe** [comment]
For what it's worth, I get a `Illegal instruction (core dumped)` error upon importing hnswlib in a docker container running on a different host than it was built on (i.e. build on Ubuntu, hosted on Debian), which is kinda weird b/c docker is supposed to flatten out differences in the host OS.  Anyway, For those in a similar situation, I use this line to test and patch in my startup script:

```bash
# ensure hnswlib is installed properly
did_succeed=$(python -c "import hnswlib; print('success',end='')")
if [ "$did_succeed" != "success" ]; then 
    pip uninstall hnswlib -y
    pip install hnswlib
    did_succeed=$(python -c "import hnswlib; print('success',end='')")
    if [ "$did_succeed" != "success" ]; then 
        echo 'hnswlib was successfully re-installed';
    else
        echo 'hnswlib re-installation FAILED';
    fi
else
    echo 'hnswlib was loaded correctly (woohoo!)';
fi
# run the app
python myapp.py
```

**Comment by yurymalkov** [comment]
@jdthorpe Hm. I think the reason for that is different instruction sets. Hnswlib is compiled with native arch, so if binaries are transferred to an older host there might be problems.
The simplest workaround is to use the least common instruction set in here https://github.com/nmslib/hnswlib/blob/d59f8d9c3932e37ed19d3893731f49dd5d09977d/setup.py#L77

**Comment by HelloWorldLTY** [comment]
Hi, would you please also fix the version in pypl? I still meet this error when using pip isntall hnswlib. Thanks!

**Comment by tleyden** [comment]
@yurymalkov I'm also seeing the error across different versions of Ubuntu.  Is that expected?  

For example if I build a container on `Ubuntu 22.04.2 LTS` and then try to run it on `Ubuntu 20.04.6 LTS`, it crashes with `Illegal instruction (core dumped)`.  

The uninstall/workaround fixes it.

**Comment by srmrz** [comment]
I can reproduce this in 0.7.0 but not in 0.8.0

Looks like this was fixed in version 0.8.0
